fili-core/src/main/java/com/yahoo/bard/webservice/web/responseprocessors/EtagCacheResponseProcessor.java

```diff
@@ -0,0 +1,134 @@
+ */
+public class EtagCacheResponseProcessor implements FullResponseProcessor {
+
+    private static final Logger LOG = LoggerFactory.getLogger(EtagCacheResponseProcessor.class);
+
+    private final ObjectMapper mapper;
+    private final ObjectWriter writer;
+    private final ResponseProcessor next;
+    private final String cacheKey;
+    private final TupleDataCache<String, String, String> dataCache;
+
+    /**
+     * Constructor.
+     *
+     * @param next  Next ResponseProcessor in the chain
+     * @param cacheKey  Key into which to write a cache entry
+     * @param dataCache  The cache into which to write a cache entry
+     * @param mapper  An object mapper to use for processing Json
+     */
+    public EtagCacheResponseProcessor(
+            @NotNull ResponseProcessor next,
+            @NotNull String cacheKey,
+            @NotNull TupleDataCache<String, String, String> dataCache,
+            @NotNull ObjectMapper mapper
+    ) {
+        this.next = next;
+        this.cacheKey = cacheKey;
+        this.dataCache = dataCache;
+        this.mapper = mapper;
+        this.writer = mapper.writer();
+    }
+
+    @Override
+    public ResponseContext getResponseContext() {
+        return next.getResponseContext();
+    }
+
+    @Override
+    public FailureCallback getFailureCallback(DruidAggregationQuery<?> druidQuery) {
+        return next.getFailureCallback(druidQuery);
+    }
+
+    @Override
+    public HttpErrorCallback getErrorCallback(DruidAggregationQuery<?> druidQuery) {
+        return next.getErrorCallback(druidQuery);
+    }
+
+    @Override
+    public void processResponse(JsonNode json, DruidAggregationQuery<?> druidQuery, LoggingContext metadata) {
+        // make sure JSON response comes with status code
+        if (!json.has(DruidJsonResponseContentKeys.STATUS_CODE.getName())) {
+            logAndGetErrorCallback(ErrorMessageFormat.STATUS_CODE_MISSING_FROM_RESPONSE.format(), druidQuery);
+            return;
+        }
+
+        int statusCode = json.get(DruidJsonResponseContentKeys.STATUS_CODE.getName()).asInt();
+        // If response is a NOT_MODIFIED, get response body from cache and inject it into JsonNode of the next
+        // response processor
+        if (statusCode == Status.NOT_MODIFIED.getStatusCode()) {
+            try {
+                ((ObjectNode) json).set(
+                        DruidJsonResponseContentKeys.RESPONSE.getName(),
+                        mapper.readTree(dataCache.getDataValue(cacheKey))
+                );
+            } catch (IOException ioe) {
+                throw new IllegalStateException(ioe);
+            }
+        } else if (statusCode == Status.OK.getStatusCode()) { // If response is a OK, cache it, including etag
+            // make sure JSON response comes with etag
+            if (!json.has(DruidJsonResponseContentKeys.ETAG.getName())) {
+                logAndGetErrorCallback(ErrorMessageFormat.ETAG_MISSING_FROM_RESPONSE.format(), druidQuery);
```

Comment:
dont throw error as commented before, instead LOG.warn and skip dataCache update and directly falls to next responseProcessor